### PR TITLE
RDBC-460 / RDBC-468 Completed tests list, implemented advanced.what_changed and session.refresh, overhauled document metadata flow

### PR DESCRIPTION
**Pull request by poissoncorp**
https://issues.hibernatingrhinos.com/issue/RDBC-460
https://issues.hibernatingrhinos.com/issue/RDBC-468

### CODE REVIEW

File: pyravendb/store/document_session.py [hunk 1]
```diff
@@ -581,6 +630,52 @@ def exists(self, key):
...
+            self.entity_changed(new_obj, self.get_document_info(document), changes)
+
+    def entity_changed(self, new_obj, document_info, changes):
+        return JsonOperation.entity_changed(
+            new_obj, document_info, changes, self.session.conventions.json_default_method
+        )
+
+    def get_metadata_for(self, entity):
+        if entity is None:
+            raise ValueError("Entity cant be None")
```
Comment (ml054) [inline pyravendb/store/document_session.py]: typo

File: pyravendb/json/json_operation.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+    @staticmethod
+    def compare_json_array(field_path, key, old_collection, new_collection, changes, doc_changes, prop_name):
+        # if we don't care about the changes
+        if len(old_collection) != len(new_collection) and changes is None:
+            return True
+        position = 0
+        changed = False
+
+        while position < len(old_collection) and position < len(new_collection):
+            if old_collection[position] is None:
```
Comment (gregolsky) [inline pyravendb/json/json_operation.py]: please extract `old_collection[position]` and `new_collection[position]` to a local var. 
https://github.com/ravendb/ravendb-nodejs-client/blob/v5.0/src/Mapping/JsonOperation.ts#L171

File: pyravendb/json/metadata_as_dictionary.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
+class MetadataAsDictionary:
+    def __init__(self, metadata=None, parent=None, parent_key=None):
+        # todo: make this condition more elegant
```
Comment (gregolsky) [inline pyravendb/json/metadata_as_dictionary.py]: what condition do you mean here?

File: pyravendb/json/metadata_as_dictionary.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
+class MetadataAsDictionary:
```
Comment (gregolsky) [inline pyravendb/json/metadata_as_dictionary.py:1]: You might want to copy these tests:
https://github.com/ravendb/ravendb-nodejs-client/blob/v5.0/test/Documents/MetadataAsDictTests.ts

File: pyravendb/store/document_session.py [hunk 5]
```diff
@@ -423,6 +430,7 @@ def counters_for(self, entity_or_document_id):
         return DocumentCounters(self, entity_or_document_id)
 
     def save_changes(self):
+        # todo: create batchOperation and call it here
```
Comment (gregolsky) [inline pyravendb/store/document_session.py]: please create a ticket for that on youtrack and ref here

File: pyravendb/store/document_session.py [hunk 6]
```diff
@@ -448,14 +483,17 @@ def _update_batch_result(self, batch_result, data):
                 if item["Type"] == "PUT":
                     entity = data.entities[i - data.deferred_command_count]
                     if entity in self._documents_by_entity:
+                        # todo: bind more info to session.documents_by_id dictionary values - like in other clients
```
Comment (gregolsky) [inline pyravendb/store/document_session.py]: ticket please - maybe list all the todos around it

File: pyravendb/store/document_session.py [hunk 7]
```diff
@@ -448,14 +483,17 @@ def _update_batch_result(self, batch_result, data):
...
                         document_info = self._documents_by_entity[entity]
+                        # todo: get_or_add_modifications
                         document_info["change_vector"] = item["@change-vector"]
                         item.pop("Type", None)
-                        document_info["original_metadata"] = item.copy()
                         document_info["metadata"] = item
                         document_info["key"] = item["@id"]
-                        document_info["original_value"] = entity.__dict__.copy()
+                        #  todo: make it more sophisticated - possibly multi-copying every reference to single object
+                        document_info["original_value"] = deepcopy(entity.__dict__)
```
Comment (gregolsky) [inline pyravendb/store/document_session.py:495]: here it's shallow copy vs deepcopy right?
doesn't it change behavior too much? Is that how java and node do here?
Also what's under `__dict__` in general?

File: pyravendb/store/document_session.py (hunk 1 above)
Comment (gregolsky) [reply pyravendb/store/document_session.py]:
```suggestion
            raise ValueError("Entity cannot be None")
```

File: pyravendb/store/entity_to_json.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
+import pyravendb.tools.utils
+from pyravendb import constants
+from copy import deepcopy
+import json
+
+
```
Comment (gregolsky) [inline pyravendb/store/entity_to_json.py:6]:
```suggestion
```

File: pyravendb/store/entity_to_json.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    def convert_entity_to_json(self, entity, document_info):
+        if document_info is not None:
+            self._session.events.before_conversion_to_json_internal(document_info["key"], entity)
+        document = EntityToJson._convert_entity_to_json_internal(self, entity, document_info)
+        if document_info is not None:
+            self._session.events.after_conversion_to_document(document_info["key"], entity, document)
+        return document
+
+    def _convert_entity_to_json_internal(self, entity, document_info, remove_identity_property=False):
+        json_node = json.loads(json.dumps(entity, default=self._session.conventions.json_default_method))
```
Comment (gregolsky) [inline pyravendb/store/entity_to_json.py]: don't you have a util wrapping dumps + loads already?

File: pyravendb/store/session_query.py [hunk 10]
```diff
@@ -1019,7 +1019,7 @@ def _execute_query(self):
         response_includes = response.pop("Includes", None)
         self.session.save_includes(response_includes)
         for result in response_results:
-            (entity, metadata, original_metadata, original_document,) = Utils.convert_to_entity(
+            (entity, metadata, original_document,) = Utils.convert_to_entity(
```
Comment (gregolsky) [inline pyravendb/store/session_query.py:1022]: what's that last left `,` for? How is it diff from https://github.com/ravendb/ravendb-python-client/pull/95/files#diff-cdc2009938de871a418ec044668392680f1374f02281db86f3251d8b38dc1c09R431

Comment (poissoncorp) [reply pyravendb/store/session_query.py:1022]: In this case, it's the same, but a comma at the end of the collection reduces diffs and including it (even inline) is considered good practice.
https://stackoverflow.com/a/7992643

File: pyravendb/store/entity_to_json.py (hunk 8 above)
Comment (poissoncorp) [reply pyravendb/store/entity_to_json.py:6]: PEP8 suggests 2 blank lines before and after class definition
https://www.python.org/dev/peps/pep-0008/#blank-lines

File: pyravendb/store/document_session.py (hunk 1 above)
Comment (poissoncorp) [reply pyravendb/store/document_session.py]: done

File: pyravendb/json/json_operation.py (hunk 2 above)
Comment (poissoncorp) [reply pyravendb/json/json_operation.py]: done

File: pyravendb/json/metadata_as_dictionary.py (hunk 3 above)
Comment (poissoncorp) [reply pyravendb/json/metadata_as_dictionary.py]: outdated todo, deleted

File: pyravendb/store/entity_to_json.py (hunk 9 above)
Comment (poissoncorp) [reply pyravendb/store/entity_to_json.py]: replaced with util

File: pyravendb/store/document_session.py (hunk 5 above)
Comment (poissoncorp) [reply pyravendb/store/document_session.py]: done

File: pyravendb/store/document_session.py (hunk 6 above)
Comment (poissoncorp) [reply pyravendb/store/document_session.py]: done

File: pyravendb/json/metadata_as_dictionary.py (hunk 4 above)
Comment (poissoncorp) [reply pyravendb/json/metadata_as_dictionary.py:1]: For sure! Pushed another commit with these

File: pyravendb/store/document_session.py (hunk 7 above)
Comment (poissoncorp) [reply pyravendb/store/document_session.py:495]: explained that it solves issues with dumping object to json string

**Pull request merged**